### ISSUE TITLE
Has hyperledger development been halted?

### ISSUE THREAD

**Issue by brianorwhatever** [hyperledger/hyperledger#64]
The last commit seems to have been in September

**Comment by christo4ferris** [comment]
Hi brianorwhatever,

Thanks for submitting this issue!

We use JIRA to track issues. we are using JIRA to track issues for the Hyperledger fabric project. You'll need a Linux ID (free) at [identity.linuxfoundation.org](https://identity.linuxfoundation.org/) and can post/track issues at [jira.hyperledger.org](http://jira.hyperledger.org). 

hyperbot

**Comment by lehors** [comment]
You should also know that this particular repository is not where code is developed. Hyperledger is an umbrella organization which hosts several projects: https://www.hyperledger.org/projects

For code, you will want to look at one of the project's repository listed on the main github page: https://github.com/hyperledger 

Beware thought that several projects, like Fabric, have moved their main repository to gerrit for its greater review capabilities and the github repo is only a mirror. See https://gerrit.hyperledger.org

**Comment by icreator** [comment]
R3 RIP

**Comment by jonschlinkert** [comment]
@lehors you could have easily used Github's API to accomplish the same thing that Gerrit does but in a much nicer user interface, and without losing the convenience, power, and collaborative features of github. it's a huge bummer that this entire project is so fragmented and the barrier to contributing is so high. 

To be honest, some of us in node.js have discussed hyperledger, and it seems that you are intentionally making it hard to contribute unless you are a blessed member of the core team. It's strange that any of the code is on GitHub at all if the intention is not to encourage collaboration.
Such a shame.

**Comment by lehors** [comment]
Hi @jonschlinkert,

I'm very sorry to hear that you have that impression. I can guarantee you that the Hyperledger developers community in general, and the Hyperledger Fabric one in particular, have no such desire. On the contrary we are constantly trying to make it easier for people to join in.

We know we have ways to go in improving documentation but just looking at the Fabric project we went from about 40 initial contributors to over 150 from 27 different organizations in about a year so I think there is evidence that we fully intend for this project to be open.

Now, I'm certainly not the one who's going to argue with you on the fact that Github's UI is better than Gerrit's, but Gerrit does allow for collaboration and is successfully used in many big open source projects. For that reason the Linux Foundation encouraged us to use Gerrit as a matter of fact and we've found Gerrit to provide us with features we wanted and which Github didn't have. Maybe we could have developed them on our own on top of Github, as you seem to suggest, but we had other priorities so we accepted the trade-off. As Github evolves we will keep an eye on it and will re-evaluate this trade-off.

In the meantime I would hope that you can pass that and join us like many others have! I sure encourage you to give it a try. :-)
Cheers.

Arnaud

**Comment by christo4ferris** [comment]
There are plenty of projects using Gerrit for code review. It isn't perfect. Nothing is, including GitHub.
 
As for your unsubstantiated claim about making it hard to contribute, care to back that up? We're all for improving the ease of contributing. 

Cheers,

Christopher Ferris
IBM Distinguished Engineer, CTO Open Technology
IBM Cloud, Open Technologies
email: chrisfer@us.ibm.com
twitter: @christo4ferris
blog: https://developer.ibm.com/opentech/author/chrisfer/
phone: +1 508 667 0402

**Comment by jonschlinkert** [comment]
> In the meantime I would hope that you can pass that and join us like many others have! I sure encourage you to give it a try. :-)

We might. I think the software itself is compelling, despite the other roadblocks to contributing. Thanks, I appreciate the feedback. 

> As for your unsubstantiated claim about making it hard to contribute, care to back that up?

Not sure where to start, @christo4ferris. I think it's pretty obvious that 1) having to leave this system to use and learn the features of another completely disconnected system, 2) not being able to comment on the code on the other system, 3) not being able to do prs through github on the other system,  and 4) not being able to take advantage of the automations, bots, and other code quality tools that we've built around our workflows for 1,400 other projects makes it harder to contribute than if the code was hosted on GitHub. If you want hard data, my projects were downloaded 640 million times last month on NPM, accounting for 6% of total downloads. How much was this downloaded?

**Comment by doowb** [comment]
@christo4ferris from my point of view, it was confusing that 17 of the 41 repositories are mirrored from Gerrit. Also, I've noticed that in some of the repositories that aren't mirrored, it's expected to file bugs and open issues in Jira (a 3rd system). I think that these are things that make it hard to contribute to the projects under hyperledger. It seems that there is an underestimation of how much developers on GitHub (perhaps more so in node.js?) rely on automated workflows and API integrations with GitHub for things like code quality. Not to mention none of the activity that happens outside of GitHub will show in a developers history. We think this is important for developers trying to build their careers and newcomers to be able to establish a track record of their contributions to the open source community.

**Comment by lehors** [comment]
@jonschlinkert It'd be helpful if you could expand a bit on the integration problem you're referring to with your workflow.

To be honest even within the Hyperledger community not everyone is happy with the switch to gerrit - Fabric started on github and as @doowb noticed not all projects have switched - so the question of switching back keeps popping back up every now and then. Any info on why it would be beneficial would be useful. It needs to be fairly detailed though so it doesn't come across as unsubstantiated claims.

@doowb regarding people's building their careers our current mirroring to github actually carries over people's activity on gerrit so that is covered.

You guys must make github people happy because you're essentially saying that to do open source it MUST be on github. Good for them bad for everyone else...

**Comment by jonschlinkert** [comment]
tldr;  I'm sorry if I sounded harsh in my prior comments. We're building a blockchain-enabled b2b supply chain platform and I think hyperledger would be a great fit for it.

> It needs to be fairly detailed though so it doesn't come across as unsubstantiated claims.

Fair enough. I have to think about how to do that. In some ways, this reminds me this question we used to ask at a prior company I worked for: "how do you quantify the value of having one mobile phone for all contacts, versus a different phone for each contact?".  I think an obvious reaction to the latter is something like: "that would just be a PITA.", and this situation is like that.  Imagine having to go to 20 different places for 20 different projects. If you're thinking, "Gerrit is just one other place". That's true. But this not the only org on GitHub. Could you imagine what it would be like if every org did this, just once? Especially considering that Github's user base is a homogenous mix of developers from everywhere around the world. Some of which work for big companies that can afford to build custom integrations and tools, but many (like @doowb and I) do not. 

At the heart of the challenge for us is the discontinuity and disconnectedness of having to use multiple systems for source control in a way that doesn't easily fit into our existing processes and workflows. It's a distraction and causes us to need to learn multiple systems. But I'm not saying this is all qualitative.  We have a number of bots and services setup around various stages of development or trigger events. Which means we'd be investing bandwidth and development time into new integrations, bots, and services for a new system with an API we're unfamiliar with, might not have the events or hooks we need, and might not ever be used again. 

I'm not saying everyone on GitHub does this, but @doowb and I do for Sellside, and many other highly effective teams do. Our observation is that devs and teams that tend to be at the "head of the long tail" tend to do these things. and the rest of the long tail benefits from it. There is a real cost to switching, but I don't know how to quantify that. 

We want to build tools around hyperledger to help build an extended ecosystem, as we've done with other ecosystems on node.js. In my experience, for us to be successful in doing this, we need to eliminate all of the technical barriers to contributing and building automated workflows. 

> You guys must make GitHub people happy because you're essentially saying that to do open source it MUST be on GitHub. Good for them bad for everyone else...

lol perhaps, but GitHub is free for FOSS projects, and there is no disadvantage to the community to have as many projects as possible in one place. The law of increasing returns says that each node added to a network increases the value of the network exponentially. This means that developers get more indirect economic value (and possibly direct economic value) when more projects are on GitHub. 

Not just developers though. Big companies like IBM, Microsoft, Facebook, and Google all stand to benefit from the large community and tremendous pool of developer talent on GitHub as well. This is why Google and Microsoft both shut down their own VCS systems in favor of joining the party on GitHub. A big win for everyone, except for GitHub's competitors. But if GitHub's competitors did a better job, we'd be having this discussion on that platform, not here. I'm happy to use a tool that works hard to make users happy, and I have no problem paying for advanced features when we need them. People often joke that programmers are anti-social. IMHO, GitHub has not only proven that wrong, they've shown the world that programmers know how to collaborate better than everyone else. 

FWIW I'm preparing a talk on "Scaling Humans" for an upcoming conference. One of the points I want to make in the talk is to "pick your battles" (which IMHO is exemplified by Google and Microsoft's decisions to drop their own VCS systems). Version control is not strategic for hyperledger, but building a developer community is (or should be if it's not).  For my part, I feel this conversation about Gerrit is worth having because I think it would be a huge win for the other 23 million developers on GitHub if hyperledger stopped hedging and dove into GitHub with both feet. We have a deep understanding of data we use to drive our decisions regarding workflows, and we've seen the impact of leveraging network effects from the developer community on GitHub. 

In case it helps to shed more perspective and lend credence to my point. @doowb and I maintain projects that are responsible for 1 in 19 downloads on NPM, and ~70 of the projects we maintain are downloaded more than 1 million times a month. Prior to programming, I was a marketing, sales and supply chain consultant for almost 20 years. We took principles that I learned during that career and applied them to our development practices. From that, we have developed a strong understanding of the formula that drives success with FOSS projects.  There is no possible way that @doowb and I alone would have the bandwidth to be able to maintain all of these code bases without finding ways to empower the community to help us, while we focus on what we're good at, and automate the rest.

**Comment by jonschlinkert** [comment]
at the risk of making a long comment even longer. I told my wife about this convo (she's also a programmer) and she suggested I add perspective on "why" I care (because she thinks I came off as a jerk in my earlier comment, and I think she's right), and she thinks maybe it will help. Here goes:

- 2010 - I joined github
- 2012 - I learned that 80% of the world popular lives on less than $2/day. I wanted to do something about it. Made my first commit, mostly CSS/HTML. I was a sales and marketing consultant for C-Level execs in supply chain, and had come up with a business model and idea for a way to overlay cloud, mobile, and social on top of existing web infrastructure to help people in impoverished regions become part of the world economy - starting by empowering small businesses in U.S. to generate revenue. But I couldn't get VC funding, idea was too new and I had no technical background. 
- 2014 - First "real code" commit. After spending two years doing design, and creating a prototype for Sellside, I learned how to program by reading code comments and literally immersing myself in code on GitHub and books (a big change in lifestyle from getting $200/hr for consulting). I wrote a long list of code tools we would need for Sellside (data translation, UI/site generating tools, etc). Every day I published one tool related to a code challenge I was learning (and was something Sellside needed). Projects started simple and got increasingly complex (and useful). Within 3 months these projects were downloaded 1m/month (my first realization of how powerful this community is). Later that year I was listed in [Open Source at Scale](https://github.com/substack/open-source-at-scale) as 8 in the top 10 contributors to open source.  @doowb joined in and we created sellside. 
- 2017 - projects created by @doowb and I are now getting 1b downloads a month. Our code and frameworks are used by WalMart, Target, Nasa, Google, Facebook, IBM, PayPal, Netflix, GE and many thousands of other teams and users. 

Recently, we learned about hyperledger and "geeked out" about how amazing and coincidental it was that this tool came out, right when we got to the point where we needed it. It overlays so perfectly on this platform we're building. We couldn't believe that it had taken so long for us to find it and were disappointed that it had been around for so long without us being able to contribute and learn. But soon afterward we realized that we can't take advantage of it anyway because we don't have the resources and bandwidth between the two of us to stretch across multiple systems (whilst also building the community we have). 

This is why I care, and I really hope we can make hyperledger a continuous part of our flow. It's all about scaling. Hope that helps.

**Comment by lehors** [comment]
Thank you for your response. This is very interesting.

For what it's worth I like the gerrit model (repo-clone) better, even though I don't like its UI (who does?). I find the triangular github model (repo-fork-clone) an unnecessary complication, but what you're saying is that this kind of consideration is not even relevant. The network effect of github trumps everything else.

As I said earlier the question of github is often discussed and I think your post will be valuable input. Thanks for the feedback.

PS: you wrote: "We're building a blockchain-enabled b2b supply chain platform and I think hyperledger would be a great fit for it." Beware that hyperledger is actually a consortium which now hosts several blockchain frameworks with different characteristics and goals. Hyperledger Fabric is one of them.

**Comment by christo4ferris** [comment]
Jon,

I'm very happy to learn that you guys care. We definitely want and need to diversify our contributors beyond what we have achieved to date. We are, as I said, open to ideas about how we can improve the ease of on-boarding new contributors (and users/consumers), as well as ease of following along with where we're going.

We plan to continue using JIRA until something better comes along and we're in a position to transition without significant disruption. Same for Gerrit. Aside from the need to obtain a LF ID (really no different than getting a GH ID) there isn't much of a barrier to engaging.

We are looking at how we can leverage JIRA better to make it clearer what's going on and where we are headed. However, every large project suffers in this regard, that it is hard to quickly grok which way is up. Whatever we can do to improve, we're all ears.

Cheers,

Christopher Ferris
IBM Distinguished Engineer, CTO Open Technology
IBM Cloud, Open Technologies
email: chrisfer@us.ibm.com
twitter: @christo4ferris
blog: https://developer.ibm.com/opentech/author/chrisfer/
phone: +1 508 667 0402

**Comment by jonschlinkert** [comment]
> Aside from the need to obtain a LF ID (really no different than getting a GH ID) there isn't much of a barrier to engaging.

I appreciate your response, but it seems clear based on your comment that you're missing the point entirely. I'm telling you that there is, in fact, a huge barrier, and I think I articulated exactly why, and it would have been more palatable to hear "my hands are tied" than for you to suggest that it's easy for us to adopt additional tools. Gerrit and JIRA do not have the same API and as GitHub, and the tools and apps that we've already invested time and dollars in for GitHub will not work with JIRA and Gerrit. When I say "we", I mean all of the 23 million GitHub users that are all here for the same reason. If you don't get that reason, then I'm probably more convinced than ever that hyperledger has ulterior motives. 

> We are looking at how we can leverage JIRA

So you're saying it's just going to keep getting worse? This is so painful and disappointing to hear. I was concerned about this project being run by IBM because of this old style thinking. So many things about the wording of the charter and the management of this project seem "fishy". If it's this much of a pain to get started, why would we invest all of our time and resources into making this awesome? You might not care because you're IBM and you don't really need us. 

Unfortunately, it looks like the best option for us is a hard fork, if we even decide hyperledger is the way to go.  Given that the code base is so fragmented, it might just be easier to create a new platform. Especially since we don't want our user base having to deal with all of this headache (and we've gotten funding recently and will continue to get more funding, so we're committed to doing this the right way).  Either way, we know how to drive adoption quickly (94%  of all node.js projects that include at least one external dependency use our code), so while it's a shame that you're digging your heels in, it might be better for everyone in the ecosystem in the long term if we start with a better, more truly open source foundation. 

If you disagree, how about if you give a reasoned argument to the rest of the community regarding why we should use those tools? You are after all using GitHub for some reason. What is that reason?

**Comment by lehors** [comment]
Jon,
You seem to keep confusing Hyperledger and Hyperledger Fabric. I don't suppose you actually mean to fork the consortium, do you? :-)

Your statement about that the code being fragmented also makes me wonder whether that's also the result of not understanding that there are different independent projects within hyperledger, I recommend you check the website to try and better understand the difference: https://www.hyperledger.org/projects

While the projects have made different choices regarding which tools they use (including whether to use Gerrit or GitHub), I believe each of them is consistent within itself.

As for the claim that this project is run by IBM and is "fishy", when it comes to Fabric it is clear that IBM plays a leading role because we contributed the code to get the project started and have more developers on it than any other organizations you seem to have a preconceived idea about the project and IBM and be looking for reasons to confirm those rather than trying to find out what really is going on.

As I indicated several times already, this is among other things a point of discussion. There are people with different point of views. Chris happens to have one you don't like but this is his opinion. I don't know why you jump on this to conclude "that hyperledger has ulterior motives". Not only this again fails to acknowledge the different projects and groups of people involved in the different hyperledger projects - including those that are on github! - it is a statement that is quite frankly unnecessarily flippant.

Cheers.

**Comment by charleshope** [comment]
What are these amazing tools that make GitHub so much better than anything else?

It's ironic that a distributed ledger project is bearing a discussion on the benefits of the extreme centralization of hosting all open source projects on with single company.

**Comment by jonschlinkert** [comment]
> You seem to keep confusing Hyperledger and Hyperledger Fabric. 

On the contrary, @doowb's comment should make it clear that we're looking at this holistically. Assuming we don't start from scratch, and we do go that route, we'll either port code that isn't in javascript, or we'll keep it synced. Of course, we'll always respect licenses and copyrights.

> it is a statement that is quite frankly unnecessarily flippant.

Sorry if it sounded that way, but there was nothing insincere or sarcastic intended in my words. Granted the comments may have been harsh, but I was being candid about how the project comes across to us. I do understand that if you're not seeing it from our perspective that it might seem like I'm being "over the top". However, if anything I think I'm holding back on how bad our initial reaction was for the sake of trying to work through this. We're both probably just seeing this through completely different lenses, with a completely different set of concerns.

> I don't suppose you actually mean to fork the consortium, do you? :-)

(As an aside, I'm not trying to "call the kettle black", but this seems like a better example of a flippant comment)

If you mean, do we plan to fork "all of the projects that have a copyleft license"? Then yes, but only it makes more sense than starting from scratch, and only the projects we need or find useful (meaning hyperledger covers other use cases than what we need). Obviously, this is not ideal. I was being forthright and 100% honest about the fact that we quite literally do not have the resources or bandwidth to incorporate other systems into our development workflow, and I was really hoping you sincerely meant that feedback might have a chance of changing aspects of the project to make it easier for devs like us. It is what it is. If you're thinking "wouldn't forking take more bandwidth than using another system?", not necessarily. One of these things is mission-critical and would allow us to justify our efforts through the promise of increasing returns, predictability, continuous integration and collaboration from the community, and ultimately more control, and the other does none of those things, and would only cost us time and money.  Since Hyperledger has a big-ish code base, forking it would not be ideal. But it's not as big as others we created and maintain and have built larger communities around than what hyperledger has today. It's our last, and probably worst option but we'll bite the bullet if we have to.

That said, by "consortium" you seem to be implying something unrelated to code, like perhaps the charter. If so, then no, I don't plan to fork the charter or other aspects of the Hyperledger organization that we either wouldn't use or don't perceive as being 100% transparent. For example, [I asked a question](https://github.com/hyperledger/hyperledger/issues/65) about premium members elsewhere 10 days ago, which still has not been answered (by a human). It's worth mentioning that that was the first time in 5 years of using GitHub, after contributing to many aspects of almost 1,700 projects that I was "redirected" to an external site by a bot, where I would be required to create a new account in another system to continue a conversation. This came across as a "bait and switch" tactic, that made this project look like a front to market to developers in an unauthentic way. I didn't mean to be flippant with my comment about it being fishy, but that's the reality of it. I perceived that, and your reaction to it seems combative instead of taking what I'm saying into consideration. 

Also, if the verbiage about "premium members" is somewhere readily available on the site or charter, or even on the hyperledger org, then my apologies. I sincerely looked for several minutes and I couldn't find it. Hopefully, you can see how that would concern anyone who wanted to invest time and effort in this project, and how I might want clarification before betting the future of a company that I have dedicated my life to building for the past 5 years. If I came across as anything other than direct about this, that wasn't my intention, and that undermines the seriousness of my point.

> What are these amazing tools that make GitHub so much better than anything else?

I'll assume this is a serious question, and not really a statement with a question mark at the end.  (As another aside, if it isn't a real question, that would be another example of "flippant"). Don't take my word for it. IBM, Google and Microsoft are much better at articulating the value. Take their word for it:

- https://siliconangle.com/blog/2016/02/23/everybody-loves-developers-github-and-ibm-announce-enterprise-collaboration-ibminterconnect/
- http://thehackernews.com/2017/04/codeplex-microsoft-github.html
- https://www.engadget.com/2015/03/13/google-code-closing/

> It's ironic that a distributed ledger project is bearing a discussion on the benefits of the extreme centralization of hosting all open source projects on with single company.

I'd be happy to discuss elsewhere since this doesn't really belong here. But distributed ledgers over blockchain solves the double spend problem. Distributes votes over blockchain solves the Byzantine Generals' Problem. GitHub is a collaboration tool that does neither of those things. Perhaps you're just making an observation about the fact that both have "nodes" and resemble networks. But in blockchains, nodes are data. In this network, nodes are people. And the value, quality, and trustworthiness of this particular network are directly related to the number and authenticity of the people using it. Consider how "proof of authority" works when the nodes are people. If that doesn't make my point, I don't know what will.

**Comment by jonschlinkert** [comment]
and @charleshope, I'll tell you what's ironic. That we, @doowb and I, a team of two developers do everything we possibly can to make our development flow easier for other developers because we understand how powerful the result of that collaboration can be. We built dashboards, slack integrations, created bots that deliver stats updates multiple times a day, daily summaries of issues and contributions from the community, and much more. Yet I'm having to debate this with IBM, a company that can easily afford to build whatever integrations and tools are necessary to make contributing easier for existing GitHub users. You are asking us to bear the cost of contributing. But if we can do it without $100 million in funding, why can't you?

**Comment by doowb** [comment]
ever seen a disappointed kid cry when they get the wrong present on Christmas? That's how @jonschlinkert and I were both feeling lol. We would both love to be able to contribute to this community (keeping fingers crossed)

**Comment by jonschlinkert** [comment]
I should mention that we're grateful to even have this conversation with you. I appreciate it.

**Comment by charleshope** [comment]
I assure you, the question was intended seriously, and not as a sarcastic comment. 

The first link is heavy on marketing buzzwords but devoid of technical considerations. The next two simply indicate that Microsoft's and Google's home grown solutions weren't as good or as popular as GitHub, and I am happy to agree with that. 

I still don't understand what these major technical differences are that result in Gerrit being virtually unusable for you. 

Tell us about your suite of awesome tools that use the GitHub API.
